halo2/src/circuit_builder.rs

```diff
@@ -0,0 +1,232 @@
+use num_bigint::BigUint;
+use polyexen::expr::{ColumnKind, Expr, PlonkVar};
+use polyexen::plaf::backends::halo2::PlafH2Circuit;
+use polyexen::plaf::{ColumnFixed, ColumnWitness, Columns, Info, Lookup, Plaf, Poly, Witness};
+
+use num_traits::One;
+use number::{BigInt, FieldElement};
+use pil_analyzer::{self, BinaryOperator, Expression, IdentityKind};
+
+use super::circuit_data::CircuitData;
+
+pub(crate) fn analyzed_to_circuit<T: FieldElement>(
+    analyzed: &pil_analyzer::Analyzed<T>,
+    fixed: Vec<(&str, Vec<T>)>,
+    witness: Vec<(&str, Vec<T>)>,
+) -> PlafH2Circuit {
+    // The structure of the table is as following
+    //
+    // | constant columns | __enable_cur | __enable_next |  witness columns | \
+    // |  c[0]            |    1         |       1       |   w[0]           |  |
+    // |  c[1]            |    1         |       1       |   w[1]           |  |>  N actual circuit rows
+    // |  ...             |   ...        |      ...      |   ...            |  |
+    // |  c[N - 2]        |    1         |       1       |   w[N - 2]       |  |
+    // |  c[N - 1]        |    1         |       0       |   w[N - 1]       | /     <- __enable_next == 0 since there's no state transition
+    // |  0               |    0         |       0       |   0              | \
+    // |  0               |    0         |       0       |   0              |  |
+    // |  ...             |   ...        |      ...      |   ...            |  |>  2**(ceil(log2(N)) + 1) - N padding rows to fit the halo2 unusable rows
```

Comment:
if the second N is the number of unusable rows, maybe rename it so it doesn't get confused with the N from "N actual circuit rows" above?

Reply:
Oh I see, added brackets to avoid confusion